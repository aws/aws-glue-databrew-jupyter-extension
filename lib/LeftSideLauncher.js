import { Widget as PhosphorWidget } from "@phosphor/widgets";
import { Widget as LuminoWidget } from "@lumino/widgets";
import { GLUE_DATABREW_RENDER } from "./constants";
export class PhosphorLeftSideLauncher extends PhosphorWidget {
    handleLaunchButtonClicked() {
        this.commands.execute(GLUE_DATABREW_RENDER);
        return;
    }
}
export class LuminoLeftSideLauncher extends LuminoWidget {
    handleLaunchButtonClicked() {
        this.commands.execute(GLUE_DATABREW_RENDER);
        return;
    }
}
export class LeftSideLauncher {
    static create(version, commands) {
        const majorVersion = Number(version.split(".")[0]);
        const widget = majorVersion === 1 ? new PhosphorLeftSideLauncher() : new LuminoLeftSideLauncher();
        widget.commands = commands;
        const launchButton = document.createElement("div");
        const header = document.createElement("header");
        header.className = "aws_glue_databrew_header";
        header.textContent = "AWS Glue DataBrew";
        launchButton.title = "Welcome to AWS Glue DataBrew";
        launchButton.id = "aws_glue_databrew_launch_button";
        launchButton.textContent = "Launch AWS Glue DataBrew";
        launchButton.onclick = widget.handleLaunchButtonClicked.bind(widget);
        widget.node.appendChild(header);
        widget.node.appendChild(launchButton);
        widget.id = "aws_glue_databrew_jupyter_left_side_launcher";
        widget.title.iconClass = "jp-databrew-logo jp-SideBar-tabIcon";
        widget.title.caption = "AWS Glue DataBrew";
        return widget;
    }
}
