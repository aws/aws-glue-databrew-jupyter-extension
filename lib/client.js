export const getAWSConfig = async (basePath) => {
    try {
        const config = await get(`${basePath}awsproxy/awsconfig`);
        return {
            region: config.region,
        };
    }
    catch (_a) {
        const config = await post(`${basePath}api/getInstanceConfig`);
        return {
            region: config.region,
        };
    }
};
export const get = (path) => {
    const options = { method: "get" };
    return request(new Request(path, options));
};
export const post = (path) => {
    const options = { method: "post" };
    return request(new Request(path, options));
};
export const request = async (req) => {
    const res = await fetch(req);
    const text = await res.text();
    const data = text && parse(text);
    if (!res.ok) {
        const error = data;
        throw new Error((error && error.message) || res.statusText);
    }
    return data;
};
export const parse = (text) => {
    try {
        return JSON.parse(text);
    }
    catch (_a) {
        return {};
    }
};
