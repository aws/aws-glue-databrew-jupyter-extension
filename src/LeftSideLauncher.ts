import { Widget as PhosphorWidget } from "@phosphor/widgets";
import { CommandRegistry as PhosphorCommandRegistry } from "@phosphor/commands";
import { Widget as LuminoWidget } from "@lumino/widgets";
import { CommandRegistry as LuminoCommandRegistry } from "@lumino/commands";
import { GLUE_DATABREW_RENDER } from "./constants";

export class PhosphorLeftSideLauncher extends PhosphorWidget {
  /**
   * Command Registry
   */
  commands: PhosphorCommandRegistry;
  handleLaunchButtonClicked(): undefined {
    this.commands.execute(GLUE_DATABREW_RENDER);
    return;
  }
}

export class LuminoLeftSideLauncher extends LuminoWidget {
  /**
   * Command Registry
   */
  commands: LuminoCommandRegistry;
  handleLaunchButtonClicked(): undefined {
    this.commands.execute(GLUE_DATABREW_RENDER);
    return;
  }
}

export class LeftSideLauncher {
  static create(version: string, commands: PhosphorCommandRegistry | LuminoCommandRegistry): PhosphorLeftSideLauncher | LuminoLeftSideLauncher {
    const majorVersion = Number(version.split(".")[0]);
    const widget = majorVersion === 1 ? new PhosphorLeftSideLauncher() : new LuminoLeftSideLauncher();
    widget.commands = commands;

    const launchButton = document.createElement("div");
    const header = document.createElement("header");
    header.className = "aws_glue_databrew_header";
    header.textContent = "AWS Glue DataBrew";

    launchButton.title = "Welcome to AWS Glue DataBrew";
    launchButton.id = "aws_glue_databrew_launch_button";
    launchButton.textContent = "Launch AWS Glue DataBrew";

    launchButton.onclick = widget.handleLaunchButtonClicked.bind(widget);

    widget.node.appendChild(header);
    widget.node.appendChild(launchButton);
    widget.id = "aws_glue_databrew_jupyter_left_side_launcher";
    widget.title.iconClass = "jp-databrew-logo jp-SideBar-tabIcon";
    widget.title.caption = "AWS Glue DataBrew";

    return widget;
  }
}
