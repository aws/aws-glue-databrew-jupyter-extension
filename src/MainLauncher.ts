import { Widget as PhosphorWidget } from "@phosphor/widgets";
import { Widget as LuminoWidget } from "@lumino/widgets";
import { AWSRegion } from "./types";
import { isCnPartition } from "./utils";
export class PhosphorMainLauncher extends PhosphorWidget {
  /**
   * The image element associated with the widget.
   */
  consoleRoot: HTMLElement;
  /**
   * A path to css assets
   */
  cssPath: string;
}

export class LuminoMainLauncher extends LuminoWidget {
  /**
   * The image element associated with the widget.
   */
  consoleRoot: HTMLElement;
  /**
   * A path to css assets
   */
  cssPath: string;
}

export class MainLauncher {
  static create(version: string, baseUrl: string, cssPath: string, region: AWSRegion): PhosphorMainLauncher | LuminoMainLauncher {
    const majorVersion = Number(version.split(".")[0]);
    const widget = majorVersion === 1 ? new PhosphorMainLauncher() : new LuminoMainLauncher();

    const awsText = isCnPartition(region) ? "Amazon" : "AWS";
    widget.cssPath = cssPath;
    widget.id = "aws_glue_databrew_jupyter";
    widget.title.label = `${awsText} Glue DataBrew`;
    widget.title.closable = true;

    widget.consoleRoot = document.createElement("html");
    widget.consoleRoot.setAttribute("style", "height: 100%");
    widget.node.appendChild(widget.consoleRoot);

    widget.consoleRoot.insertAdjacentHTML("beforeend",
      `
        <head>
          <meta charset="UTF-8">
          <meta name="awsc-lang" content="en">
          <meta id="jupyter-server-base-url" content="${baseUrl}">
          <meta id="jupyterlab-version" content="${version}">
          <meta name="aws-glue-databrew-jupyter" content="true">
          <meta id="aws-glue-databrew-jupyter-region" content="${region || ""}">
          <title>AWS</title>
          <link rel="stylesheet" href="${cssPath}">
          <style>
            .loader {
              display: inline-block;
              border: 5px solid #f3f3f3;
              border-radius: 50%;
              border-top: 5px solid #3498db;
              width: 40px;
              height: 40px;
              margin-right: 12px;
              -webkit-animation: spin 2s linear infinite;
              animation: spin 2s linear infinite;
            }

            @-webkit-keyframes spin {
              0% { -webkit-transform: rotate(0deg); }
              100% { -webkit-transform: rotate(360deg); }
            }

            @keyframes spin {
              0% { transform: rotate(0deg); }
              100% { transform: rotate(360deg); }
            }

            .loader-container {
              font-size: 60px;
              font-weight: 100;
              color: #879196;
              height: 100%;
              width: 100%;
              justify-content: center;
              display: flex;
              align-items: center;
            }
          </style>
        </head>
        <body style="height: 100%">
          <div id="app" style="height: 100%">
            <div class="loader-container">
              <div>
                <div class="loader"></div><span>Launching ${awsText} Glue DataBrew</span>
              </div>
            </div>
          </div>
        </body>
    `
    );

    return widget;
  }
}
