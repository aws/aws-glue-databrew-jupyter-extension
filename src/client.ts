import { AWSConfig } from "./types";

export const getAWSConfig = async (basePath: string): Promise<AWSConfig> => {
  try {
    const config = await get<AWSConfig>(`${basePath}awsproxy/awsconfig`);
    return {
      region: config.region,
    };
  } catch {
    const config = await post<AWSConfig>(`${basePath}api/getInstanceConfig`);
    return {
      region: config.region,
    };
  }
};

export const get = <T>(path: string): Promise<T> => {
  const options = { method: "get" };
  return request(new Request(path, options));
};

export const post = <T>(path: string): Promise<T> => {
  const options = { method: "post" };
  return request(new Request(path, options));
};

export const request = async <T>(req: Request): Promise<T> => {
  const res = await fetch(req);
  const text = await res.text();
  const data = text && parse<T | Error>(text);
  if (!res.ok) {
    const error = data as Error;
    throw new Error((error && error.message) || res.statusText);
  }
  return data as T;
};

export const parse = <T>(text: string): T => {
  try {
    return JSON.parse(text);
  } catch {
    return {} as T;
  }
};
