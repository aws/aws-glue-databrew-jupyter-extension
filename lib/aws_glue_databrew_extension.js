import { ILayoutRestorer, } from '@jupyterlab/application';
import { ICommandPalette, } from '@jupyterlab/apputils';
import { ILauncher, } from '@jupyterlab/launcher';
import { GLUE_DATABREW_RENDER } from './constants';
import { LeftSideLauncher, } from './LeftSideLauncher';
import { MainLauncher, } from './MainLauncher';
const getAppVersion = (app) => Number(app.version.split('.')[0]);
/**
 * Initialize the console widget extension
 */
export const initiateExtension = (getPaths) => {
    const activate = async (app, palette, restorer, launcher) => {
        const version = getAppVersion(app);
        // Declare a widget variable
        const [jsPath, cssPath] = await getPaths();
        const consoleWidget = MainLauncher.create(version, cssPath);
        app.commands.addCommand(GLUE_DATABREW_RENDER, {
            label: 'Launch AWS Glue DataBrew',
            icon: 'jp-databrew-logo',
            execute: () => {
                if (!consoleWidget.isAttached) {
                    app.shell.add(consoleWidget, 'main');
                    const script = document.createElement('script');
                    script.setAttribute('src', jsPath);
                    consoleWidget.consoleRoot.appendChild(script);
                }
                app.shell.activateById(consoleWidget.id);
            },
        });
        const launcherWidget = LeftSideLauncher.create(version, app.commands);
        restorer.add(launcherWidget, launcherWidget.id);
        app.shell.add(launcherWidget, 'left');
        // Add the command to the palette.
        palette.addItem({ command: GLUE_DATABREW_RENDER, category: 'Launcher' });
        if (launcher) {
            const launcher_item = {
                command: GLUE_DATABREW_RENDER,
                args: {
                    newBrowserTab: true,
                    title: 'Launch Databrew',
                    id: 'databrew-launcher'
                },
                category: 'Other'
            };
            launcher.add(launcher_item);
        }
    };
    const extension = {
        id: 'aws_glue_databrew_jupyter',
        autoStart: true,
        requires: [ICommandPalette, ILayoutRestorer, ILauncher],
        activate: activate,
    };
    return extension;
};
