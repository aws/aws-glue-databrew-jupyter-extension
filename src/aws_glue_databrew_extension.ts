import {
  JupyterFrontEnd,
  JupyterFrontEndPlugin,
  ILayoutRestorer,
} from "@jupyterlab/application";
import { ICommandPalette } from "@jupyterlab/apputils";
import { ILauncher } from "@jupyterlab/launcher";
import { Widget } from "@lumino/widgets";
import { getAWSConfig } from "./client";
import { GLUE_DATABREW_RENDER } from "./constants";
import { LeftSideLauncher } from "./LeftSideLauncher";
import { MainLauncher } from "./MainLauncher";
import { isCnPartition } from "./utils";

const getAppVersion = (app: JupyterFrontEnd) => app.version;
const getBaseUrl = (app: JupyterFrontEnd) =>
  app.serviceManager.serverSettings.baseUrl;

/**
 * Initialize the console widget extension
 */
export const initiateExtension = (
  getPaths: (region: string) => Promise<string[]>
): JupyterFrontEndPlugin<void> => {
  const activate = async (
    app: JupyterFrontEnd,
    palette: ICommandPalette,
    restorer: ILayoutRestorer,
    launcher: ILauncher
  ) => {
    const version = getAppVersion(app);
    const baseUrl = getBaseUrl(app);
    const url = new URL(baseUrl);
    const { region } = await getAWSConfig(url.pathname);
    const [jsPath, cssPath] = await getPaths(region);
    const awsText = isCnPartition(region) ? "Amazon" : "AWS";
    const consoleWidget = MainLauncher.create(
      version,
      baseUrl,
      cssPath,
      region
    );

    app.commands.addCommand(GLUE_DATABREW_RENDER, {
      label: `Launch ${awsText} Glue DataBrew`,
      icon: "jp-databrew-logo",
      execute: () => {
        if (!consoleWidget.isAttached) {
          app.shell.add(consoleWidget as Widget, "main");
          const script = document.createElement("script");

          script.setAttribute("src", jsPath);
          consoleWidget.consoleRoot.appendChild(script);
        }
        app.shell.activateById(consoleWidget.id);
      },
    });

    const launcherWidget = LeftSideLauncher.create(version, app.commands);

    restorer.add(launcherWidget as Widget, launcherWidget.id);
    app.shell.add(launcherWidget as Widget, "left");

    // Add the command to the palette.
    palette.addItem({ command: GLUE_DATABREW_RENDER, category: "Launcher" });
    if (launcher) {
      const launcher_item: ILauncher.IItemOptions = {
        command: GLUE_DATABREW_RENDER,
        args: {
          newBrowserTab: true,
          title: "Launch Databrew",
          id: "databrew-launcher",
        },
        category: "Other",
      };

      launcher.add(launcher_item);
    }
  };

  const extension: JupyterFrontEndPlugin<void> = {
    id: "aws_glue_databrew_jupyter",
    autoStart: true,
    requires: [ICommandPalette, ILayoutRestorer, ILauncher],
    activate: activate,
  };

  return extension;
};
