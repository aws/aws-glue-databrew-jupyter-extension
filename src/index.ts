import { initiateExtension } from "./aws_glue_databrew_extension";
import {
  getContentPrefixUrl,
  getPrefixUrl,
} from "./utils";
import { jsFileName, styleFileName } from "./constants";

export const getPaths = async (region: string): Promise<string[]> => {
  const prefix = await getPrefix(region);
  const contentPrefixUrl = `${getContentPrefixUrl(region)}/${prefix}`;
  return [
    `${contentPrefixUrl}/${jsFileName}`,
    `${contentPrefixUrl}/${styleFileName}`,
  ];
};

export const getPrefix = async (region: string): Promise<string> => {
  const prefixUrl = getPrefixUrl(region);
  const req = new Request(prefixUrl);
  const response = await fetch(req);
  const json = await response.json();
  return json && json.prefix.toString();
};

const extension = initiateExtension(getPaths);

export default extension;
