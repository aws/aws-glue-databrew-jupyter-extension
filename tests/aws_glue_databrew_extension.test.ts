import { ILayoutRestorer, JupyterFrontEnd } from "@jupyterlab/application";
import { ServiceManager, ServerConnection } from "@jupyterlab/services";
import { ICommandPalette } from "@jupyterlab/apputils";
import { ILauncher } from "@jupyterlab/launcher";
import { CommandRegistry } from "@lumino/commands";
import * as plugin from "../src/aws_glue_databrew_extension";
import { LuminoMainLauncher, MainLauncher } from "../src/MainLauncher";
import { LeftSideLauncher, LuminoLeftSideLauncher } from "../src/LeftSideLauncher";
import * as client from "../src/client";

describe("initiateExtension", () => {
  test("should return extension object", () => {
    const getPaths = jest.fn().mockReturnValueOnce(Promise.resolve(["main.js", "styles.css"]));
    const extension = plugin.initiateExtension(getPaths);
    expect(extension.id).toBe("aws_glue_databrew_jupyter");
    expect(extension.autoStart).toBe(true);
    expect(extension.requires).toEqual([ICommandPalette, ILayoutRestorer, ILauncher]);
    expect(extension.activate).toEqual(expect.any(Function));
  });

  describe("activate", () => {
    test("should activate plugin", async () => {
      const getPaths = jest.fn().mockReturnValueOnce(Promise.resolve(["main.js", "styles.css"]));
      jest.spyOn(MainLauncher, "create").mockReturnValueOnce(new LuminoMainLauncher());
      jest.spyOn(LeftSideLauncher, "create").mockReturnValueOnce(new LuminoLeftSideLauncher());
      jest.spyOn(client, "getAWSConfig").mockReturnValueOnce(Promise.resolve({
        region: "us-east-1",
      }));
      const extension = plugin.initiateExtension(getPaths);
      const commands: Partial<CommandRegistry> = {
        addCommand: jest.fn(),
      };
      const serviceManager = {
        serverSettings: {
          baseUrl: "http://localhost:8888/",
        } as ServerConnection.ISettings,
      } as ServiceManager;
      const shell: Partial<JupyterFrontEnd.IShell> = {
        add: jest.fn(),
        activateById: jest.fn(),
      };
      const app: Partial<JupyterFrontEnd>  = {
        commands: commands as CommandRegistry,
        shell: shell as JupyterFrontEnd.IShell,
        serviceManager,
        version: "1.0.0",
      };
      const palette: Partial<ICommandPalette> = {
        addItem: jest.fn(),
      };
      const restorer: Partial<ILayoutRestorer> = {
        add: jest.fn(),
      };
      const launcher: Partial<ILauncher> = {
        add: jest.fn(),
      };
      await extension.activate(
        app as JupyterFrontEnd,
        palette,
        restorer,
        launcher,
      );
      expect(getPaths).toHaveBeenCalledTimes(1);
      expect(commands.addCommand).toHaveBeenCalledTimes(1);
      expect(palette.addItem).toHaveBeenCalledTimes(1);
      expect(shell.add).toHaveBeenCalledTimes(1);
      expect(launcher.add).toHaveBeenCalledTimes(1);
    });

    test("should activate plugin in cn region", async () => {
      const getPaths = jest.fn().mockReturnValueOnce(Promise.resolve(["main.js", "styles.css"]));
      jest.spyOn(MainLauncher, "create").mockReturnValueOnce(new LuminoMainLauncher());
      jest.spyOn(LeftSideLauncher, "create").mockReturnValueOnce(new LuminoLeftSideLauncher());
      jest.spyOn(client, "getAWSConfig").mockReturnValueOnce(Promise.resolve({
        region: "cn-north-1",
      }));
      const extension = plugin.initiateExtension(getPaths);
      const commands: Partial<CommandRegistry> = {
        addCommand: jest.fn(),
      };
      const serviceManager = {
        serverSettings: {
          baseUrl: "http://localhost:8888/",
        } as ServerConnection.ISettings,
      } as ServiceManager;
      const shell: Partial<JupyterFrontEnd.IShell> = {
        add: jest.fn(),
        activateById: jest.fn(),
      };
      const app: Partial<JupyterFrontEnd>  = {
        commands: commands as CommandRegistry,
        shell: shell as JupyterFrontEnd.IShell,
        serviceManager,
        version: "1.0.0",
      };
      const palette: Partial<ICommandPalette> = {
        addItem: jest.fn(),
      };
      const restorer: Partial<ILayoutRestorer> = {
        add: jest.fn(),
      };
      const launcher: Partial<ILauncher> = {
        add: jest.fn(),
      };
      await extension.activate(
        app as JupyterFrontEnd,
        palette,
        restorer,
        launcher,
      );
      expect(getPaths).toHaveBeenCalledTimes(1);
      expect(commands.addCommand).toHaveBeenCalledTimes(1);
      expect(palette.addItem).toHaveBeenCalledTimes(1);
      expect(shell.add).toHaveBeenCalledTimes(1);
      expect(launcher.add).toHaveBeenCalledTimes(1);
    });

    test("should activate plugin without launcher", async () => {
      const getPaths = jest.fn().mockReturnValueOnce(Promise.resolve(["main.js", "styles.css"]));
      jest.spyOn(MainLauncher, "create").mockReturnValueOnce(new LuminoMainLauncher());
      jest.spyOn(LeftSideLauncher, "create").mockReturnValueOnce(new LuminoLeftSideLauncher());
      jest.spyOn(client, "getAWSConfig").mockReturnValueOnce(Promise.resolve({
        region: "us-east-1",
      }));
      const extension = plugin.initiateExtension(getPaths);
      const commands: Partial<CommandRegistry> = {
        addCommand: jest.fn(),
      };
      const serviceManager = {
        serverSettings: {
          baseUrl: "http://localhost:8888/",
        } as ServerConnection.ISettings,
      } as ServiceManager;
      const shell: Partial<JupyterFrontEnd.IShell> = {
        add: jest.fn(),
        activateById: jest.fn(),
      };
      const app: Partial<JupyterFrontEnd>  = {
        commands: commands as CommandRegistry,
        shell: shell as JupyterFrontEnd.IShell,
        serviceManager,
        version: "1.0.0",
      };
      const palette: Partial<ICommandPalette> = {
        addItem: jest.fn(),
      };
      const restorer: Partial<ILayoutRestorer> = {
        add: jest.fn(),
      };
      const launcher: Partial<ILauncher> = {
        add: jest.fn(),
      };
      await extension.activate(
        app as JupyterFrontEnd,
        palette,
        restorer,
        null,
      );
      expect(getPaths).toHaveBeenCalledTimes(1);
      expect(commands.addCommand).toHaveBeenCalledTimes(1);
      expect(palette.addItem).toHaveBeenCalledTimes(1);
      expect(shell.add).toHaveBeenCalledTimes(1);
      expect(launcher.add).toHaveBeenCalledTimes(0);
    });
  });

  describe("execute", () => {
    test("should handle unattached widget", async () => {
      const getPaths = jest.fn().mockReturnValueOnce(Promise.resolve(["main.js", "styles.css"]));
      const widget = new LuminoMainLauncher();
      widget.consoleRoot = document.createElement("html");
      const mainLauncherMock = jest.spyOn(MainLauncher, "create").mockReturnValueOnce(widget);
      jest.spyOn(LeftSideLauncher, "create").mockReturnValueOnce(new LuminoLeftSideLauncher());
      jest.spyOn(client, "getAWSConfig").mockReturnValueOnce(Promise.resolve({
        region: "us-east-1",
      }));
      const extension = plugin.initiateExtension(getPaths);
      const commands: Partial<CommandRegistry> = {
        addCommand: jest.fn().mockImplementationOnce((_: string, options: { label: string, icon: string, execute: () => void }) => {
          options.execute();
          expect(widget.consoleRoot).toMatchSnapshot();
        }),
      };
      const serviceManager = {
        serverSettings: {
          baseUrl: "http://localhost:8888/",
        } as ServerConnection.ISettings,
      } as ServiceManager;
      const shell: Partial<JupyterFrontEnd.IShell> = {
        add: jest.fn(),
        activateById: jest.fn(),
      };
      const app: Partial<JupyterFrontEnd>  = {
        commands: commands as CommandRegistry,
        shell: shell as JupyterFrontEnd.IShell,
        serviceManager,
        version: "1.0.0",
      };
      const palette: Partial<ICommandPalette> = {
        addItem: jest.fn(),
      };
      const restorer: Partial<ILayoutRestorer> = {
        add: jest.fn(),
      };
      const launcher: Partial<ILauncher> = {
        add: jest.fn(),
      };
      await extension.activate(
        app as JupyterFrontEnd,
        palette,
        restorer,
        launcher,
      );
      expect(getPaths).toHaveBeenCalledTimes(1);
      expect(commands.addCommand).toHaveBeenCalledTimes(1);
      expect(palette.addItem).toHaveBeenCalledTimes(1);
      expect(shell.add).toHaveBeenCalledTimes(2);
      expect(launcher.add).toHaveBeenCalledTimes(1);
      expect(mainLauncherMock).toHaveBeenCalledTimes(1);
    });

    test("should handle attached widget", async () => {
      const getPaths = jest.fn().mockReturnValueOnce(Promise.resolve(["main.js", "styles.css"]));
      const widget = new LuminoMainLauncher();
      const mainLauncherMock = jest.spyOn(MainLauncher, "create").mockImplementationOnce(() => {
        Object.defineProperty(widget, "isAttached", { value: true });
        widget.id = "id";
        return widget;
      });
      jest.spyOn(LeftSideLauncher, "create").mockReturnValueOnce(new LuminoLeftSideLauncher());
      jest.spyOn(client, "getAWSConfig").mockReturnValueOnce(Promise.resolve({
        region: "us-east-1",
      }));
      const extension = plugin.initiateExtension(getPaths);
      const commands: Partial<CommandRegistry> = {
        addCommand: jest.fn().mockImplementationOnce((_: string, options: { label: string, icon: string, execute: () => void }) => {
          options.execute();
          expect(widget.consoleRoot).toMatchSnapshot();
        }),
      };
      const serviceManager = {
        serverSettings: {
          baseUrl: "http://localhost:8888/",
        } as ServerConnection.ISettings,
      } as ServiceManager;
      const shell: Partial<JupyterFrontEnd.IShell> = {
        add: jest.fn(),
        activateById: jest.fn(),
      };
      const app: Partial<JupyterFrontEnd>  = {
        commands: commands as CommandRegistry,
        shell: shell as JupyterFrontEnd.IShell,
        serviceManager,
        version: "1.0.0",
      };
      const palette: Partial<ICommandPalette> = {
        addItem: jest.fn(),
      };
      const restorer: Partial<ILayoutRestorer> = {
        add: jest.fn(),
      };
      const launcher: Partial<ILauncher> = {
        add: jest.fn(),
      };
      await extension.activate(
        app as JupyterFrontEnd,
        palette,
        restorer,
        launcher,
      );
      expect(getPaths).toHaveBeenCalledTimes(1);
      expect(commands.addCommand).toHaveBeenCalledTimes(1);
      expect(palette.addItem).toHaveBeenCalledTimes(1);
      expect(shell.add).toHaveBeenCalledTimes(1);
      expect(shell.activateById).toHaveBeenCalledTimes(1);
      expect(shell.activateById).toHaveBeenCalledWith("id");
      expect(launcher.add).toHaveBeenCalledTimes(1);
      expect(mainLauncherMock).toHaveBeenCalledTimes(1);
    });
  });
});
