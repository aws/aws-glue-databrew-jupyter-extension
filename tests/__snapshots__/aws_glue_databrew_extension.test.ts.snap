// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initiateExtension execute should handle attached widget 1`] = `undefined`;

exports[`initiateExtension execute should handle unattached widget 1`] = `
<html>
  <script
    src="main.js"
  />
</html>
`;
